File: egosync-app/src-tauri/src/models/skill.rs
pub const BUTLER_SCOPE_ID: &str = "__butler__";
pub const SOURCE_TYPE_CUSTOM: &str = "custom";
pub const SOURCE_TYPE_OPENCODE: &str = "opencode";

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SkillRoleScope {
    #[serde(default)]
    pub all_roles: bool,
    #[serde(default)]
    pub role_ids: Vec<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, sqlx::FromRow, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SkillRegistryEntry {
    pub id: String,
    pub name: String,
    pub description: String,
    pub source_type: String,
    pub managed_path: String,
    pub content_hash: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SkillImportPreview {
    pub name: String,
    pub description: String,
    pub content_hash: String,
    pub duplicate: Option<SkillDuplicateInfo>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct SkillDuplicateInfo {
    pub kind: String,
    pub existing: SkillRegistryEntry,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct OpencodeSkillCandidate {
    pub name: String,
    pub description: String,
    pub source_location: String,
    pub source_path: String,
    pub source_type: String,
    pub content_hash: String,
    pub already_imported: bool,
    pub duplicate: Option<SkillDuplicateInfo>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct OpencodeSkillSkippedSummary {
    pub total: usize,
    pub reasons: Vec<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DiscoverOpencodeSkillsResult {
    pub items: Vec<OpencodeSkillCandidate>,
    pub skipped: OpencodeSkillSkippedSummary,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportOpencodeSkillInput {
    pub source_path: String,
    #[serde(default)]
    pub role_scope: Option<SkillRoleScope>,
    /// 发现时展示给用户的 content_hash。导入时用于比对源文件是否在
    /// discover→import 之间被替换（TOCTOU 一致性校验）。为兼容旧调用允许缺省。
    #[serde(default)]
    pub expected_content_hash: Option<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ImportOpencodeSkillResult {
    pub status: String,
    pub entry: Option<SkillRegistryEntry>,
    /// registry 写入后是否已成功同步到 opencode agent 配置。
    /// command 层在 full_sync 失败时置为 false，前端据此避免谎称"已启用"。
    pub synced: bool,
    pub runtime_ready: bool,
    pub runtime_error: Option<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct PickCustomSkillDirectoryResult {
    pub content: String,
    pub source_path: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PreviewCustomSkillInput {
    pub content: Option<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportCustomSkillInput {
    pub content: Option<String>,
    #[serde(default)]
    pub source_path: Option<String>,
    #[serde(default)]
    pub overwrite_existing: bool,
    #[serde(default)]
    pub role_scope: Option<SkillRoleScope>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ImportCustomSkillResult {
    pub status: String,
    pub entry: Option<SkillRegistryEntry>,
    pub preview: SkillImportPreview,
    pub runtime_ready: bool,
    pub runtime_error: Option<String>,
}

File: egosync-app/src/types/skill.ts
export interface SkillRoleScope {
  allRoles: boolean;
  roleIds: string[];
}

export interface SkillRegistryEntry {
  id: string;
  name: string;
  description: string;
  sourceType: 'custom' | 'opencode';
  managedPath: string;
  contentHash: string;
  createdAt: string;
  updatedAt: string;
}


export type SelectableSkillKey = `registry:${string}` | 'meta:find-skills' | 'meta:skill-creator';

export interface SelectableSkill {
  key: SelectableSkillKey;
  name: string;
  description: string;
  kind: 'registry' | 'meta';
  sourceType: 'custom' | 'opencode' | 'meta';
}

export interface SkillScopeUpdatedPayload {
  scopeKind: 'role' | 'butler' | 'all';
  ownerId: string | null;
}

export interface SkillDuplicateInfo {
  kind: 'contentHash' | 'name';
  existing: SkillRegistryEntry;
}

export interface OpencodeSkillCandidate {
  name: string;
  description: string;
  sourceLocation: string;
  sourcePath: string;
  sourceType: 'opencode';
  contentHash: string;
  alreadyImported: boolean;
  duplicate: SkillDuplicateInfo | null;
}

export interface OpencodeSkillSkippedSummary {
  total: number;
  reasons: string[];
}

export interface DiscoverOpencodeSkillsResult {
  items: OpencodeSkillCandidate[];
  skipped: OpencodeSkillSkippedSummary;
}

export interface ImportOpencodeSkillInput {
  sourcePath: string;
  roleScope?: SkillRoleScope;
  /** 发现时展示的 content_hash，导入时回传供后端做 TOCTOU 一致性校验。 */
  expectedContentHash?: string;
}

export interface ImportOpencodeSkillResult {
  status: 'imported' | 'duplicate';
  entry: SkillRegistryEntry | null;
  /** registry 写入后是否已成功同步到 opencode agent。false 时不应声称"已启用"。 */
  synced: boolean;
  runtimeReady: boolean;
  runtimeError: string | null;
}

export interface SkillImportPreview {
  name: string;
  description: string;
  contentHash: string;
  duplicate: SkillDuplicateInfo | null;
}

export interface PickCustomSkillDirectoryResult {
  content: string;
  sourcePath: string;
}

export interface PreviewCustomSkillInput {
  content?: string;
}

export interface ImportCustomSkillInput {
  content?: string;
  sourcePath?: string;
  overwriteExisting?: boolean;
  roleScope?: SkillRoleScope;
}

export interface ImportCustomSkillResult {
  status: 'imported' | 'duplicate';
  entry: SkillRegistryEntry | null;
  preview: SkillImportPreview;
  runtimeReady: boolean;
  runtimeError: string | null;
}
